### PR TITLE
Add transactions to rayyan-store endpoints

### LINKED ISSUE #21: Wrap queries in certain rayyan-store endpoints in a transaction
The SQL queries in the following endpoints need to be executed in a transaction:

```
- deleteByFileId
- insertDedup
- updateCustomization (Not 100% sure since it is a stored function)
```

Explore [this](https://github.com/vitaly-t/pg-promise#transactions) library for transactions on top of `node-pg`.

### PR DESCRIPTION
**Pull request by fabubaker**
* `executeAsTransaction(pool, queries)` has been implemented in a separate module to help run queries in a transaction. It returns the results of the last query in the transaction.

* `deleteByFileId` & `insertDedup` have been modified to use the new transaction interface.

Fixes #21

### CODE REVIEW

File: server/src/adapters/pg-transaction.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
+const _ = require('lodash')
```
Comment (hammady) [inline server/src/adapters/pg-transaction.js]: Please encapsulate in a new `postgres` folder as described [here](https://nodejs.org/api/modules.html#modules_folders_as_modules).

File: server/src/adapters/pg-transaction.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  executeAsTransaction: function (pool, queries) {
+  // @brief Runs a chain of queries in the order provided
+  // in a transaction.
+  // @param {Pool} pool - node-pg pool object
+  // @param {Array} queries - Array of query objects to chain in a transaction.
+  //                          Objects must be the query string
+  //                          OR
+  //                          of the form {text: <query string>, values: <parameters>}
+    if (queries.length === 0) {
+      // Why?!
```
Comment (hammady) [inline server/src/adapters/pg-transaction.js]: ?

File: server/src/adapters/pg-transaction.js [hunk 3]
```diff
@@ -10,65 +10,53 @@ module.exports = {
...
-    })
+
+        return chain
+          .then(function (lastResult) {
+            console.log('[TRANSACTION] Committing')
+            return client.query('COMMIT')
+              .then(function () {
+                client.release()
+                // return last query result
+                return Promise.resolve(lastResult)
```
Comment (hammady) [inline server/src/adapters/pg-transaction.js]: Just return `lastResult`

**Comment by fabubaker** [comment]
I believe [this](https://stackoverflow.com/questions/12778209/are-postgresql-functions-transactional]) answers the question of whether Postgres function are transactional.

File: server/src/adapters/postgres/pg-transaction.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+const _ = require('lodash')
```
Comment (hammady) [inline server/src/adapters/postgres/pg-transaction.js]: This declaration pollutes the global scope, please move inside the module.

File: server/src/adapters/postgres/pg-transaction.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  //                          of the form {text: <query string>, values: <parameters>}
+    if (queries.length === 0) {
+      return Promise.resolve()
+    }
+
+    if (queries.length === 1) {
+      return pool.query(queries[0])
+    }
+
+    console.log('[TRANSACTION] Connecting to client')
```
Comment (hammady) [inline server/src/adapters/postgres/pg-transaction.js]: I'm concerned about too many console logs that will fill up the logs. Can we control its level?

File: server/src/adapters/postgres/pg-transaction.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+              .then(function () {
+                client.release()
+                // return last query result
+                return lastResult
+              })
+          })
+          .catch(function (error) {
+            console.log('[TRANSACTION] Error with transaction, rolling back')
+            return client.query('ROLLBACK')
+              .catch(function () {
```
Comment (hammady) [inline server/src/adapters/postgres/pg-transaction.js]: Need to receive the error argument in this catch to report it back.

File: server/src/adapters/postgres/pg-transaction.js (hunk 5 above)
Comment (fabubaker) [reply server/src/adapters/postgres/pg-transaction.js]: I'll open an issue to add logging functionality to rayyan-store.

**Pull request merged**